datadog_test.go

```diff
@@ -0,0 +1,294 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package datadog
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"log"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+var (
+	measureCount = stats.Int64("fooCount", "testing count metrics", stats.UnitBytes)
+	measureSum   = stats.Int64("fooSum", "testing sum metrics", stats.UnitBytes)
+	measureLast  = stats.Int64("fooLast", "testing LastValueData metrics", stats.UnitBytes)
+	measureDist  = stats.Int64("fooHisto", "testing histogram metrics", stats.UnitDimensionless)
+	testTags     []tag.Key
+)
+
+func newView(agg *view.Aggregation) *view.View {
+	return &view.View{
+		Name:        "fooCount",
+		Description: "fooDesc",
+		Measure:     measureCount,
+		Aggregation: agg,
+	}
+}
+
+func customNewView(measureName string, agg *view.Aggregation, tags []tag.Key, measure *stats.Int64Measure) *view.View {
+	return &view.View{
+		Name:        measureName,
+		Description: "fooDesc",
+		Measure:     measureCount,
+		TagKeys:     tags,
+		Aggregation: agg,
+	}
+}
+
+func TestExportView(t *testing.T) {
+	exporter := NewExporter(Options{})
+
+	view.RegisterExporter(exporter)
+	reportPeriod := time.Millisecond
+	view.SetReportingPeriod(reportPeriod)
+
+	vd := &view.Data{
+		View: customNewView("fooCount", view.Count(), testTags, measureCount),
+	}
+	if err := view.Register(vd.View); err != nil {
+		t.Fatalf("Register error occurred: %v\n", err)
+	}
+	defer view.Unregister(vd.View)
+	// Wait for exporter to process metrics
+	<-time.After(10 * reportPeriod)
+
+	ctx := context.Background()
+	stats.Record(ctx, measureCount.M(1))
+	<-time.After(10 * time.Millisecond)
+
+	actual := exporter.collector.viewData["fooCount"].View
+	if actual != vd.View {
+		t.Errorf("Expected: %v, Got: %v\n", vd, actual)
+	}
+}
+
+func TestSanitizeString(t *testing.T) {
+	str1 := sanitizeString("data-234_123!doge")
```

Comment:
Since this test actually contains two tests, let's make use of [sub-tests](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks). You can see some more examples in the blog post at [https://blog.golang.org/subtests](https://blog.golang.org/subtests).